poetry/puzzle/solver.py

```diff
@@ -417,6 +420,12 @@ def __init__(
     def reachable(self) -> List["PackageNode"]:
         children: List[PackageNode] = []
 
+        # skip already traversed packages
+        if self.package in self.seen:
```

Comment:
Just an idea: could/should this check be moved to https://github.com/python-poetry/poetry/pull/3839/files#diff-c2fd13b8785fbb20817a8c0e4ce7425249eaa6be7a4faeb1fad846cadccfbfadR363 (part of the depth-first-search method implementation)?

Something along the lines of:

```python
    ...
    for neighbor in node.reachable():
        back_edges[neighbor.id].append(node)
        if neighbor.id in visited:
            continue
        ...
```

(it feels like the check is an optimization of the DFS algorithm, which is why I wondered if it could live closer to the `dfs_visit` method rather than being attached to the nodes that the search runs over.  I haven't performance evaluated this though, nor checked whether it would behave the same way as your change)

Reply:
@jayaddison quite possibly!  I kept the fix as close to the original as possible from August 2020 (before the DFS was more clearly formalized in the code).  I expect that the regular authors/maintainers who are more familiar with the internal structure and flow will know where to best plug this fix in. 🙂